contract_comm/types.go

```diff
@@ -0,0 +1,135 @@
+	MakeCall(registryId common.Hash, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error)
+}
+
+// Creates a new EVM on demand.
+type evmBuilder interface {
+	// createEVM creates an EVM. May fail due to `chain.State()`
+	createEVM() (*vm.EVM, error)
+}
+
+// Private struct that implements SystemContractCaller.
+type contractCommunicator struct {
+	builder evmBuilder
+}
+
+var emptyMessage = types.NewMessage(common.HexToAddress("0x0"), nil, 0, common.Big0, 0, common.Big0, nil, nil, common.Big0, []byte{}, false, false)
+
+// specificStateCaller implements evmBuilder for a given state. This allows core contract calls against and arbitrary state
+type specificStateCaller struct {
+	header *types.Header
+	state  vm.StateDB
+	chain  vm.ChainContext
+}
+
+func (c specificStateCaller) createEVM() (*vm.EVM, error) {
+	if internalEvmHandlerSingleton == nil || c.chain == nil {
+		return nil, ccerrors.ErrNoInternalEvmHandlerSingleton
+	}
+	context := vm.NewEVMContext(emptyMessage, c.header, c.chain, nil)
+	return vm.NewEVM(context, c.state, c.chain.Config(), *c.chain.GetVMConfig()), nil
+}
+
+// TODO(Joshua): Inject chain here instead of using singleton
+// currentStateCaller creates an EVM from the current blockchain state
+type currentStateCaller struct{}
+
+var internalEvmHandlerSingleton *InternalEVMHandler
+
+// An EVM handler to make calls to smart contracts from within geth
+type InternalEVMHandler struct {
+	chain vm.ChainContext
+}
+
+// Sets the singletone. Currently required. To eliminate must pass install of contract caller to each contract_comm method
+func SetInternalEVMHandler(chain vm.ChainContext) {
+	if internalEvmHandlerSingleton == nil {
+		log.Trace("Setting the InternalEVMHandler Singleton")
+		internalEvmHandler := InternalEVMHandler{
+			chain: chain,
+		}
+		internalEvmHandlerSingleton = &internalEvmHandler
+	}
+}
+
+func (c currentStateCaller) createEVM() (*vm.EVM, error) {
+	header := internalEvmHandlerSingleton.chain.CurrentHeader()
+	var state vm.StateDB
+	state, err := internalEvmHandlerSingleton.chain.State()
+	if err != nil {
+		log.Error("Error in retrieving the state from the blockchain", "err", err)
+		return nil, err
+	}
+
+	// The EVM Context requires a msg, but the actual field values don't really matter for this case.
+	// Putting in zero values.
+	context := vm.NewEVMContext(emptyMessage, header, internalEvmHandlerSingleton.chain, nil)
+	return vm.NewEVM(context, state, internalEvmHandlerSingleton.chain.Config(), *internalEvmHandlerSingleton.chain.GetVMConfig()), nil
+}
+
+// NewCaller creates a caller object that acts on the supplied statedb in the environment of the header, state
+// SetIEVMHandler must be called prior to using this (to maintain backwards compatibility)
+func NewCaller(header *types.Header, state vm.StateDB) (SystemContractCaller, error) {
```

Comment:
Related to my comment above, I'd recommend moving `NewCaller` and `NewCurrentStateCaller` into `caller.go`